rand_distr/src/dirichlet.rs

```diff
@@ -9,12 +9,182 @@
 
 //! The dirichlet distribution.
 #![cfg(feature = "alloc")]
-use num_traits::Float;
-use crate::{Distribution, Exp1, Gamma, Open01, StandardNormal};
-use rand::Rng;
+use crate::{Beta, Distribution, Exp1, Gamma, Open01, StandardNormal};
 use core::fmt;
-#[cfg(feature = "serde_with")]
-use serde_with::serde_as;
+use num_traits::{Float, NumCast};
+use rand::Rng;
+#[cfg(feature = "serde_with")] use serde_with::serde_as;
+
+use alloc::{boxed::Box, vec, vec::Vec};
+
+#[derive(Clone, Debug, PartialEq)]
+#[cfg_attr(feature = "serde_with", serde_as)]
+struct DirichletFromGamma<F, const N: usize>
+where
+    F: Float,
+    StandardNormal: Distribution<F>,
+    Exp1: Distribution<F>,
+    Open01: Distribution<F>,
+{
+    samplers: [Gamma<F>; N],
+}
+
+/// Error type returned from `DirchletFromGamma::new`.
+#[cfg_attr(doc_cfg, doc(cfg(feature = "alloc")))]
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+enum DirichletFromGammaError {
+    /// Gamma::new(a, 1) failed.
+    GammmaNewFailed,
+
+    /// gamma_dists.try_into() failed (in theory, this should not happen).
+    GammaArrayCreationFailed,
+}
```

Comment:
I'm not 100% sure about the error types here. Ideally, they should refer to invalid parameters supplied by the caller.

However, in this case they are only used internally, so I think this is fine.